### PR TITLE
Use pathlib for path manipulations

### LINKED ISSUE #100: It's 2018, let's use `pathlib`
`pathlib` is cool.  Should update the tools to use it.

### PR DESCRIPTION
**Pull request by mdboom**
Fixes #100.

### CODE REVIEW

File: benchmark/benchmark.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 import re
 import subprocess
 import sys
 
-sys.path.insert(0, os.path.abspath(
-     os.path.join(os.path.dirname(__file__), '..', 'test')))
+sys.path.insert(
+    0, str((Path(__file__).resolve().parent.parent / 'test')))
+
+print(sys.path)
```
Comment (rth) [inline benchmark/benchmark.py]: Forgotten print entry?

File: test/make_test_list.py [hunk 2]
```diff
@@ -3,21 +3,24 @@
...
         root = ''
     else:
         root = '.'.join(root.split('/')) + '.'
 
     for filename in files:
-        if filename.startswith("test_") and filename.endswith(".py"):
-            tests.append(root + os.path.splitext(filename)[0])
+        filename = Path(filename)
+        if str(filename).startswith("test_") and filename.suffix == ".py":
+            tests.append(root / filename.stem)
```
Comment (rth) [inline test/make_test_list.py]: Maybe better to cast `str(root / filename.stem)` explicitly here?

I guess `fp.write(test + '\n')` will do it, but not sure, also not sure how sorting works with pathlib objects..

File: test/test_python.py [hunk 3]
```diff
@@ -295,8 +295,7 @@ def pytest_generate_tests(metafunc):
         test_modules = []
         if 'CIRCLECI' not in os.environ or True:
             with open(
-                    str(pathlib.Path(__file__).parents[0] /
-                        "python_tests.txt")) as fp:
+                    Path(__file__).parents[0] / "python_tests.txt") as fp:
```
Comment (rth) [inline test/test_python.py]: Isn't `Path.parents[0]` equivalent to `Path.parent` ?

File: benchmark/benchmark.py [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
-import os
+from pathlib import Path
 import re
 import subprocess
 import sys
 
-sys.path.insert(0, os.path.abspath(
-     os.path.join(os.path.dirname(__file__), '..', 'test')))
+sys.path.insert(
+    0, str((Path(__file__).resolve().parent.parent / 'test')))
```
Comment (rth) [inline benchmark/benchmark.py]: Minor comment `.parent.parent` should be equivalent to`.parents[1]` I think. Not sure which one is better.

File: tools/buildall [hunk 5]
```diff
@@ -78,8 +75,9 @@ def parse_args():
 
 
 def main(args):
-    packagesdir = os.path.abspath(args.dir[0])
-    build_packages(packagesdir, args)
+    packagesdir = Path(args.dir[0]).resolve()
+    outputdir = Path(args.output[0]).resolve()
+    build_packages(packagesdir, outputdir)
```
Comment (rth) [inline tools/buildall]: It seems to ignore all the other `args`  from `parse_args` except for `dir` and `output` (e.g.  `cflags`, `ldflags` etc) : is this something we want to do?

File: tools/common.py [hunk 6]
```diff
@@ -1,11 +1,9 @@
...
 
 
-ROOTDIR = os.path.abspath(os.path.dirname(__file__))
-HOSTPYTHON = os.path.abspath(
-    os.path.join(ROOTDIR, '..', 'cpython', 'build', '3.6.4', 'host'))
-TARGETPYTHON = os.path.abspath(
-    os.path.join(ROOTDIR, '..', 'cpython', 'installs', 'python-3.6.4'))
+ROOTDIR = Path(__file__).parent.resolve()
+HOSTPYTHON = ROOTDIR / '..' / 'cpython' / 'build' / '3.6.4' / 'host'
+TARGETPYTHON = ROOTDIR / '..' / 'cpython' / 'installs', 'python-3.6.4'
```
Comment (rth) [inline tools/common.py]: `'installs' / 'python-3.6.4'`

**Comment by mdboom** [comment]
Merging, since all of the failures are instances of #90.

**Pull request merged**